### PR TITLE
Refactored & Added a Script to be Run Locally for a Project

### PR DESCRIPTION
**Pull request by nicky-johnstone**
Testing
---

```
tclsh all.tcl 
Tests running in interp:  /usr/bin/tclsh
Tests located in:  /home/nick/ci-development/tests/tcl
Tests running in:  /home/nick/ci-development/tests/tcl
Temporary files stored in /home/nick/ci-development/tests/tcl
Test files run in separate interpreters
Running tests that match:  *
Skipping test files that match:  l.*.test
Only running test files that match:  *.test
Tests began at Thu Mar 02 18:35:25 GMT 2023
linter.test
test_coverage.test

Tests ended at Thu Mar 02 18:35:25 GMT 2023
all.tcl:	Total	24	Passed	24	Skipped	0	Failed	0
Sourced 2 Test Files.
```

### CODE REVIEW

**Comment by nicky-johnstone** [comment]
Updated with optional named args and support for a config file.

All args are optional in either the command line or config file.

Command line args take priority over the config file args.

If `tcl_dir` and `test_dir` are not specified in either the command line or the config file then the current working directory and all subdirectories will be searched for Tcl and test files.

The config file can exist in either the current working directory or in the home directory. Current working directory takes priority over the home directory.

Example usage:

```shell
cd ~/ci-development
tclsh ~/ci-development/scripts/linting.tcl -line_length 100 -file_length 500 -proc_length 80 -tcl_dir tcl -test_dir tests

tclsh ~/ci-development/scripts/linting.tcl
```

Example config files:
```
file_length = 1000
proc_length = 100
line_length = 90
tcl_dir = tcl
test_dir = test
```

```
file_length = 400
proc_length = 50
```

```
test_dir = tests
```

**Comment by nicky-johnstone** [comment]
Example output:

```
tclsh ~/ci-development/scripts/linting.tcl -line_length 85
Checking for files that are more than 400 lines long.
---

0 files exceeding 400 lines found.

--------------------------------------------------

Checking line length is under 85 characters.
---
Line 79 in file /home/nick/ci-development/tcl/linter.tcl is 88 chars long - error $message [...
Line 143 in file /home/nick/ci-development/tcl/linter.tcl is 88 chars long - error $message [...
Line 272 in file /home/nick/ci-development/tcl/linter.tcl is 88 chars long - error $message [...
Line 136 in file /home/nick/ci-development/tcl/test_coverage.tcl is 89 chars long - #| i.e. test GET...
Line 273 in file /home/nick/ci-development/tcl/test_coverage.tcl is 87 chars long - dict set data pr...
Line 276 in file /home/nick/ci-development/tcl/test_coverage.tcl is 90 chars long - dict set data en...
Line 329 in file /home/nick/ci-development/tcl/test_coverage.tcl is 86 chars long - puts "The proc \...

7 lines exceeding 85 characters found.

--------------------------------------------------

Checking for procs that have bodies that are more than 80 lines long.
---

0 procs with bodies exceeding 80 lines found.

--------------------------------------------------

Checking for proc names that are not prefixed with the file name.
---

0 procs without file name as a prefix found.

--------------------------------------------------

Checking for procs that do not have at least one unit test.
---

0 procs that do not have a unit test found.
```

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	mkdir $(TEMP_PATH)
+	curl --fail -K ~/.curlrc_github -L -o v$(VERSION).tar.gz \
+	https://api.github.com/repos/qcode-software/ci-development/tarball/v$(VERSION)
+	tar --strip-components=1 --exclude Makefile --exclude .github --exclude tcl \
+	--exclude tests -xzvf v$(VERSION).tar.gz -C $(TEMP_PATH)
+	# checkinstall
+	fakeroot checkinstall -D --deldoc --backup=no --install=no --pkgname=$(DPKG_NAME) \
+	--pkgversion=$(VERSION) --pkgrelease=$(RELEASE) -A all -y \
+	--maintainer $(MAINTAINER) --pkglicense="BSD" --reset-uids=yes \
+	--requires "tcl,tcllib,html2text,curl,tclcurl" --replaces none --conflicts none \
```
Comment (bvw) [inline Makefile:26]: tcl and tcllib should be enough ?

**Comment by nicky-johnstone** [comment]
I've updated the Makefile so that a Tcl package called `qcode-linter` is created and installed alongside the qcode-linter script being set up.

I've tested the Makefile in a limited capacity on my dev box by running some of the shell commands. A full test of the Makefile will only be possible once the linting script is in the master branch and a release has been created.

**Comment by nicky-johnstone** [comment]
Counts have been removed from the report procs.

### Testing
- [x] Run each script via command line.
- [x] Run parts of Makefile to install qcode-linter script and qcode-linter Tcl package.
- [x] Run qcode-linter on this project.

#### Unit Tests
```
tclsh all.tcl
Tests running in interp:  /usr/bin/tclsh
Tests located in:  /home/nick/ci-development/tests/tcl
Tests running in:  /home/nick/ci-development/tests/tcl
Temporary files stored in /home/nick/ci-development/tests/tcl
Test files run in separate interpreters
Running tests that match:  *
Skipping test files that match:  l.*.test
Only running test files that match:  *.test
Tests began at Tue Mar 07 14:27:24 GMT 2023
linter.test
test_coverage.test

Tests ended at Tue Mar 07 14:27:24 GMT 2023
all.tcl:	Total	24	Passed	24	Skipped	0	Failed	0
Sourced 2 Test Files.
```

File: scripts/linting.tcl [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+puts ""
+puts "--------------------------------------------------"
+puts ""
+puts "Checking line length is under $line_length characters."
+puts "---"
+
+set long_lines [linter_lines_over_length $tcl_files $line_length]
+linter_report_lines_over_length $long_lines
+set count 0
+foreach value [dict values $long_lines] {
```
[the remark this answers is not in the document]
Comment (bvw) [reply scripts/linting.tcl:149]:
```tcl
proc linter_report_lines_over_length {
				      files
				      max_line_length
				  } {
    #| Report on lines in files longer than max line length.
    set report [list]
    foreach file $files {
	set contents [cat $file]
	set line_no 1

	foreach line [split $contents "\n"] {
	    set line_length [string length $line]

	    if { $line_length > $max_line_length } {
		set partial_line [string range [string trimleft $line] 0 15]
		lappend report "Line $line_no in file $filename is $line_length chars long"
		lappend report " - ${partial_line}..."        
            }
            incr line_no
        }
    }

    return [join $report \n]
}

proc linter_check_lines_over_length {
				     files
				     max_line_length
				 } {
    #| Return a count of lines in files longer than max line length.
    set count 0 
    foreach file $files {
	set contents [cat $file]
	
	foreach line [split $contents "\n"] {
	    if { [string length $line] > $max_line_length } {
		incr count
            }
        }
    }
    return $count
}

```

**Comment by nicky-johnstone** [comment]
### Testing
- [x] Run each script via command line.
- [x] Run parts of Makefile to install qcode-linter script and qcode-linter Tcl package.
- [x] Run qcode-linter on this project.

#### Unit Tests

```
Tests running in interp:  /usr/bin/tclsh
Tests located in:  /home/nick/ci-development/tests/tcl
Tests running in:  /home/nick/ci-development/tests/tcl
Temporary files stored in /home/nick/ci-development/tests/tcl
Test files run in separate interpreters
Running tests that match:  *
Skipping test files that match:  l.*.test
Only running test files that match:  *.test
Tests began at Wed Mar 08 18:16:08 GMT 2023
linter.test
test_coverage.test

Tests ended at Wed Mar 08 18:16:08 GMT 2023
all.tcl:	Total	25	Passed	25	Skipped	0	Failed	0
Sourced 2 Test Files.
```

**Pull request merged**